sklearn/preprocessing/_polynomial.py

```diff
@@ -497,12 +419,26 @@ def transform(self, X):
                 # edge case: deal with empty matrix
                 XP = sparse.csr_matrix((n_samples, 0), dtype=X.dtype)
             else:
-                # scipy.sparse.hstack breaks in scipy<1.9.2
+                # `scipy.sparse.hstack` breaks in scipy<1.9.2
                 # when `n_output_features_ > max_int32`
-                # TODO: Remove _csr_hstack and instead use
-                # scipy.sparse.hstack when we support
-                # scipy>=1.9.2.
-                XP = _csr_hstack(to_stack, dtype=X.dtype)
+                all_int32 = True
+                for mat in to_stack:
+                    all_int32 &= mat.indices.dtype == np.int32
+                if (
+                    sp_version < parse_version("1.9.2")
+                    and self.n_output_features_ > max_int32
+                    and all_int32
+                ):
+                    raise ValueError(
+                        "Due to an error in `scipy.sparse.hstack` present in versions"
+                        " `<1.9.2`, stacking sparse matrices such that the resulting"
+                        " matrix would have `n_cols` too large to be represented by"
+                        " 32bit integers results in negative columns. To avoid this"
+                        " error, either use a version of scipy `>=1.9.2` or alter the"
+                        " `PolynomialFeatures` transformer to produce fewer output"
+                        " features."
```

Comment:
A suggestion to mention that using `np.int64` for `indices` and `indptr` should work
```suggestion
                        "Due to a bug in `scipy.sparse.hstack` present in SciPy<1.9.2,"
                        " stacking sparse matrices such that the resulting matrix would"
                        " have its `n_cols` too large to be represented by 32bit
                        " integers results in negative column indices.\n"
                        " To avoid this error, either use `scipy>=1.9.2`, convert your"
                        " input matrix `indices` and `indptr` array to use `np.int64`
                        " instead of `np.int32`, or alter the `PolynomialFeatures`
                        " transformer to produce fewer output features."
```

Reply:
> A suggestion to mention that using np.int64 for indices and indptr should work

Actually, I'm not sure that would work due to the eager downcasting of  CSR matrices upon construction. If all the `indices, indptr` of the CSR matrices to be stacked are `int64` but _could be_ `int32` then the `hstack` function will downcast them, so it only works if, for example, they have a value large enough to _require_ `int64`.

Reply:
Ah yes, I missed the downcast in `_stack_along_minor_axis` which is called in `bmat`. You can ignore my comment.